rewrite-java-17/src/main/java/org/openrewrite/java/isolated/ReloadableJava17ParserVisitor.java

```diff
@@ -1308,10 +1284,55 @@ public J visitTypeCast(TypeCastTree node, Space fmt) {
 
     @Override
     public J visitAnnotatedType(AnnotatedTypeTree node, Space fmt) {
+        if (node.getUnderlyingType() instanceof ArrayTypeTree) {
+            return visitAnnotatedArrayType(node, fmt);
+        }
         return new J.AnnotatedType(randomId(), fmt, Markers.EMPTY, convertAll(node.getAnnotations()),
                 convert(node.getUnderlyingType()));
     }
 
+    private J visitAnnotatedArrayType(Tree node, Space fmt) {
+        if (!(node instanceof ArrayTypeTree) && !(node instanceof AnnotatedTypeTree)) {
+            throw new IllegalArgumentException("Unexpected Tree subtype " + node.getClass().getName());
+        }
+        Tree currentNode = node;
+        while (currentNode instanceof ArrayTypeTree || currentNode instanceof AnnotatedTypeTree) {
+            if (currentNode instanceof ArrayTypeTree) {
+                currentNode = ((ArrayTypeTree) currentNode).getType();
+            } else {
+                currentNode = ((AnnotatedTypeTree) currentNode).getUnderlyingType();
+            }
+        }
+
+        TypeTree elemType = convert(currentNode);
+
+        List<J.ArrayType.AnnotatedDimension> dimensions = new ArrayList<>();
+        currentNode = node;
+        while (currentNode instanceof ArrayTypeTree || currentNode instanceof AnnotatedTypeTree) {
```

Comment:
It seems the while loop here covers the while loop above (lines 1299-1305),  which looks redundant to me, since `elemType` is the only result we want from the above while loop and it can be fetched from this loop as well.

Reply:
Since `convert` is responsible for scanning the whitespace between tokens using an internal cursor that is incremented left-to-right, it's essential that `elemType` (which contains the type identifier of the elements in the array type, e.g. the `int` in `int @NotNull []` in line 1307 is converted first since it appears first in the array type.

To collapse this into one while loop, I would have to store the collected annotations (line 1314) and dimensions (line 1319) so that I can call `convertAll` and `sourceBefore` after `elemType` is converted; maybe through something like

```
dimensions = collectedAnnotations.map(anns -> new AnnotatedDimension(
        convertAll(anns),
        padRight(sourceBefore("["), sourceBefore("]"))
    );
```

How does that sound?

Reply:
I see, thanks for the explanation, I think both look good to me, seems the 2nd approach (collect everything first)  looks better.